homeassistant/components/sensor/history_stats.py

```diff
@@ -0,0 +1,304 @@
+"""
+Component to make instant statistics about your history.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.history_stats/
+"""
+
+import asyncio
+import logging
+import math
+import re
+import time
+
+import voluptuous as vol
+
+import homeassistant.components.history as history
+import homeassistant.components.recorder as recorder
+import homeassistant.helpers.config_validation as cv
+import homeassistant.util.dt as dt_util
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (CONF_NAME, CONF_ENTITY_ID, CONF_STATE)
+from homeassistant.core import callback
+from homeassistant.exceptions import TemplateError
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_state_change
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'history_stats'
+DEPENDENCIES = ['history']
+
+CONF_START = 'start'
+CONF_END = 'end'
+CONF_DURATION = 'duration'
+CONF_PERIOD_KEYS = [CONF_START, CONF_END, CONF_DURATION]
+
+DEFAULT_NAME = 'unnamed statistics'
+UNIT = 'h'
+UNIT_RATIO = '%'
+ICON = 'mdi:chart-line'
+
+PRINT_START = 'from'
+PRINT_END = 'to'
+PRINT_VALUE = 'value'
+PRINT_RATIO = 'ratio'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_ENTITY_ID): cv.entity_id,
+    vol.Required(CONF_STATE): cv.slug,
+    vol.Optional(CONF_START, default=None): cv.template,
+    vol.Optional(CONF_END, default=None): cv.template,
+    vol.Optional(CONF_DURATION, default=None): cv.template,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+
+# noinspection PyUnusedLocal
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the History Stats sensor."""
+    entity_id = config.get(CONF_ENTITY_ID)
+    entity_state = config.get(CONF_STATE)
+    start = config.get(CONF_START)
+    end = config.get(CONF_END)
+    duration = config.get(CONF_DURATION)
+    name = config.get(CONF_NAME)
+    hsh = HistoryStatsHelper
+
+    params = 0
+    for param in [start, end, duration]:
+        if param is not None:
+            params += 1
+
+    if params != 2:
+        _LOGGER.error('You must provide exactly 2 of the following:'
+                      + ' start, end, duration')
+        return False
```

Comment:
`vol.Inclusive` is good when you want either 3/3 or 0/3 parameters.
`vol.Exclusive` is good when you want either 1/3 or 0/3 parameters.
What I want is exactly 2/3 of them, not more, not less. That means my schema can be one of 3 possible schemas.

I have been trying to do combinations of these for the last 2 hours but I really cannot make it, I'm really confused by how voluptuous works, I'm sorry.